DATAES-799 - Support optimistic locking for full update scenario using seq_no + primary_term

rpuch opened the pull request:
There is a couple of issues that need resolving. Please provide your opinion on them.

1. `SeqNoPrimaryTerm` property is not written to a `Document` nor it is read from a `Document` during the mapping; it is 'implicitly transient', even if the corresponding field is marked with a `@Field` annotation. But if someone marks such a property with a `@Field` annotation, it will be added to a mapping. This looks like a (minor) discrepancy.
**a**. Should we leave it as it is now?
**b**. Or, if such a field is marked as a `@Field`, should we both map it (to/from a `Document`) and add it it a mapping?
**c**. Or should we avoid reading/writing/adding it to a mapping in any case, even if it is explicitly annotated with `@Field`?
2. `ElasticsearchTemplate` did not use any exception translator. As a result, Elasticsearch-specific exceptions are thrown whenever they occur. It seems logical to throw an `OptimisticLockingFailureException` if a seq_no-related conflict occurs, so I added such a conversion in just one method (indexing one). We need to decide whether it is ok to use an exception translator there (maybe it was not used there by design?), and, if yes, whether it should be used everywhere where `ActionFuture` result is obtained. My suggestion for now is to just use it in one indexing method, but expand its usage in a subsequent PR dedicated to exception translation in `ElasticsearchTemplate` as this PR is already pretty large. Anyway, I added a TODO.
3. if_seq_no and version are not allowed by Elasticsearch in the same index request, it just returns an error for such a request. It may happen that an entity class has both `SeqNoPrimaryTerm` and `@Version` properties. In such a case, we could de one of the following:
a. Throw an exception meaning that such a mapping is invalid
b. Only send if_seq_no and **do not** send version. Here, the `@Version` property only works in a 'read-only' mode: it allows to peek at what version was assigned by the Elasticsearch cluster to a document.
I've implemented option **b** for now, because it seems that the use-case with 'read-only' version field may be useful. But, at the same time, a programmer may be confused by his 'version-based optimistic locking' being silently disabled, so option **a** also seems viable.

<!--

Thank you for proposing a pull request. This template will guide you through the essential steps necessary for a pull request.
Make sure that:

-->

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATAES).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

sothawo commented:
Just had a short glimpse, I will have a close look the next days

As for your questions:

1. _answer c)_. I don't think that these values should be mapped in the the __source_ of an elasticsearch document. When an entity is written that has a `SeqNoPrimaryTerm` these values are only used for constructing the request parameters and flags. If it were added to the document and then written to the __source_, the value in elasticsearch that is in the
 document is different from what will be returned for __seq_no_ and __primary_term_ because on the writing elasticsearch increased the values. So it must not be written to the mappings as well. The `MappingBuilder` must check if a property is of type `SeqNoPrimaryTerm` and if so, skip it. We might consider logging a warning if the property has a `@Field
 ` annotation.

2. Having the `ElasticsearchExceptionTranslator` is a pretty recent addition, that we only added to the rest client (the reactive part had it before) after we changed all the calls to go through the `execute(callback)` method. We did not do this on purpose on the transport client up to now, as we hope that Elasticsearch 8 will be out soon, so we can get
 rid of the transport client then, and we try to keep the work on that part low. So use for this case as you do, but I don't think we need to rework this in the whole transport client.

3. Not sure about that. In the Elasticsearch documentation about optimistic looking, there is no mention of the __version_ field. I would prefer _answer b_, but with an additional warning logged.

rpuch commented:
1. I've implemented option c) (under no circumstances SeqNoPrimaryTerm property gets mapped or put into mapping). A warning is logged if such a property is annotated with `@Field` or a similar annotation.
2. Usage of an exception translator is left only in 'index' scenario for `ElasticsearchTemplate`.
3. Option b) is implemented. A warning is logged if an entity is both versioned and contains SeqNoPrimaryTerm property.

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/document/MapDocument.java:
```diff
@@ -114,6 +118,68 @@ public void setVersion(long version) {
+	 */
+	@Override
+	public boolean hasSeqNo() {
+		return this.seqNo != null && SequenceNumbers.isAssignedSeqNo(this.seqNo);
```
```suggestion
		return this.seqNo != null;
```

I wouldn't do the validity check here but at the place where a `SeqNoPrimaryTerm`is constructed. Here in the `MapDocument` these are just numeric values coming from ES.

rpuch replied:
Done

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/document/MapDocument.java:
```diff
@@ -114,6 +118,68 @@ public void setVersion(long version) {
+	 */
+	@Override
+	public boolean hasPrimaryTerm() {
+		return this.primaryTerm != null && SequenceNumbers.isAssignedPrimaryTerm(this.primaryTerm);
```
```suggestion
		return this.primaryTerm != null;
```

same as above

rpuch replied:
Done

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/convert/MappingElasticsearchConverter.java:
```diff
@@ -220,6 +218,17 @@ public void afterPropertiesSet() {
 
 	}
 
+	private SeqNoPrimaryTerm fillSeqNoPrimaryTermFromDocument(Document document) {
+		SeqNoPrimaryTerm seqNoPrimaryTerm = new SeqNoPrimaryTerm();
+		if (document.hasSeqNo()) {
+			seqNoPrimaryTerm.setSequenceNumber(document.getSeqNo());
+		}
+		if (document.hasPrimaryTerm()) {
+			seqNoPrimaryTerm.setPrimaryTerm(document.getPrimaryTerm());
+		}
+		return seqNoPrimaryTerm;
+	}
```
would not be needed, when the factory method `SeqNoPrimaryTerm.of(...)` is used like above

rpuch replied:
This logic has been moved to `SeqNoPrimaryTerm.ofAssigned()` which returns an `Optional`.

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/SequenceNumbers.java:
```diff
@@ -0,0 +1,32 @@
+ * @author Roman Puchkovskiy
+ * @since 4.0
+ */
+public class SequenceNumbers {
```
I don't think we need this class at all if we change `SeqNoPrimaryTerm`to an immutable data class and put these checks in the factory method.

rpuch replied:
The class has been removed

rpuch commented on src/main/java/org/springframework/data/elasticsearch/core/mapping/SeqNoPrimaryTerm.java:
```diff
@@ -0,0 +1,77 @@
+ * @author Roman Puchkovskiy
+ * @since 4.0
+ */
+public class SeqNoPrimaryTerm {
```
@sothawo An immutable value object makes a lot of sense, but I was keeping the following in mind. Imagine that this feature is used in a web-application. A browser makes a request to a RESTful API and gets a response with the data of the entity for editing. Such data must include seqNo and primaryTerm, because they will need to be sent back to the server upon entity save process when the user is done editing. This means that there should be a way to reconstruct the `SeqNoPrimaryTerm` instance from the data sent from the client. Also, it would be great if such a reconstruction could be done automatically from, let's say, JSON, so that we do not force a programmer to reconstruct it manually.

A mutable structure makes it easy. But having this class an immutable value object, we force a programmer to call `SeqNoPrimaryTerm.of()` manually in some kind of a converter.

Is this ok?

rpuch replied:
I've transformed `SeqNoPrimaryTerm` to an immutable value object, but the problem of its usage in a JSON DTO remains open and, I believe, needs an additional discussion.

rpuch commented on src/main/java/org/springframework/data/elasticsearch/core/mapping/SeqNoPrimaryTerm.java:
```diff
@@ -0,0 +1,77 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.elasticsearch.core.mapping;
```
Moved the class to the `org.springframework.data.elasticsearch.core.query` package

rpuch commented on src/main/java/org/springframework/data/elasticsearch/core/mapping/ElasticsearchPersistentEntity.java:
```diff
@@ -96,4 +97,34 @@
+	@Nullable
+	ElasticsearchPersistentProperty getSeqNoPrimaryTermProperty();
+	
+	default ElasticsearchPersistentProperty getRequiredSeqNoPrimaryTermProperty() {
```
Done

rpuch commented on src/main/java/org/springframework/data/elasticsearch/core/mapping/SimpleElasticsearchPersistentEntity.java:
```diff
@@ -231,6 +237,30 @@ public void addPersistentProperty(ElasticsearchPersistentProperty property) {
+			if (hasSeqNoPrimaryTermProperty()) {
+				logger.warn("Both SeqNoPrimaryTerm and @Version properties are defined on {}. Version will not be sent in index requests when seq_no is sent!", getType());
+			}
+		}
```
If we just apply the suggested change, the warning will be emitted for each property added after both `SeqNoPrimaryTerm` property and a versioned property have been added, so there can be a lot of such warnings per entity class. I've made the check twice for the 2 cases: when `SeqNoPrimaryTerm` property is added before and after the version property.

I've extracted the duplicated logging to a common private method. Is it enough? If not, please elaborate.

rpuch commented on src/main/java/org/springframework/data/elasticsearch/core/index/MappingBuilder.java:
```diff
@@ -167,6 +167,13 @@ private void mapEntity(XContentBuilder builder, @Nullable ElasticsearchPersisten
 					}
 
+					if (property.isSeqNoPrimaryTermProperty()) {
+						logger.warn("Property {} of {} is annotated for inclusion in mapping, but its type is " + //
```
Done

sothawo replied:
that's skipping the property when it is of the correct type **and** it is annotated with `@Field`; skipping must be done as well when there is no `@Field`

rpuch replied:
Oops, my bad. Fixed, this time correctly, I hope :)

rpuch commented:
Thank you for the review.

I've resolved (or commented on) all the issues. Please let me know if something still needs to be changed.

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/convert/MappingElasticsearchConverter.java:
```diff
@@ -205,6 +197,13 @@ public void afterPropertiesSet() {
 				}
 			}
+
+			if (targetEntity.hasSeqNoPrimaryTermProperty() && document.hasSeqNo() && document.hasPrimaryTerm()) {
+				SeqNoPrimaryTerm.ofAssigned(document.getSeqNo(), document.getPrimaryTerm()).ifPresent(seqNoPrimaryTerm -> {
+					ElasticsearchPersistentProperty property = targetEntity.getRequiredSeqNoPrimaryTermProperty();
+					targetEntity.getPropertyAccessor(result).setProperty(property, seqNoPrimaryTerm);
+				});
+			}
```
```suggestion
			if (targetEntity.hasSeqNoPrimaryTermProperty() && document.hasSeqNo() && document.hasPrimaryTerm()) {
				SeqNoPrimaryTerm seqNoPrimaryTerm = new SeqNoPrimaryTerm(document.getSeqNo(), document.getPrimaryTerm());
					ElasticsearchPersistentProperty property = targetEntity.getRequiredSeqNoPrimaryTermProperty();
					targetEntity.getPropertyAccessor(result).setProperty(property, seqNoPrimaryTerm);
			}
```
if the `SeqNoPrimaryTerm` constructor is added as I wrote there

rpuch replied:
Elasticsearch can actually return 'unassigned' values for seq_no and primary_term. Earlier you suggested to return them 'as is', just as numbers, without any interpretation, from `Document.getSeqNo()` and `Document.getPrimaryTerm()`, so the analysis of their 'validity' was moved to the 'optional factory' method, namely `SeqNoPrimaryTerm.ofAssigned()`. But if we remove that factory method and just invoke a constructor that blows up on an 'unassigned' value, it will throw an exception in such cases.

I suggest to do the following:
1. Make the constructor public to allow Jackson use it for instance construction
2. Still leave `ofAssigned()` and use it here (in the mapping code) to sort out the cases when Elasticsearch returns unassigned values
3. Add assertions to the constructor to avoid garbage going from some kind of a REST API.

What do you think?

sothawo replied:
Where is this documented? Then I'd suggest in `MappingElasticsearchConverter:202` to check the values of `document.getSeqNo()` and `document.getPrimaryTerm()` with a simple `if` check and set the `SeqNoPrimaryTerm` if these are correct. This is the only place where this check is used.

As https://www.elastic.co/guide/en/elasticsearch/reference/current/optimistic-concurrency-control.html does not talk about potentially unassigned values I  think it might be confusing to introduce this into the code of `SeqNoPrimaryTerm`.

We still need the `IllegalArgumentException in case of the user setting some bad values.

rpuch replied:
The `search` API does not return `seq_no` and `primary_term` by default, you need to specifically ask it to return the real values (using `seq_no_primary_term` field), otherwise it returns 'unassigned' values. We send `seq_no_primary_term=true` when doing searches, but it seems a bit fragile to hope that unassigned values will never be returned by Elasticserch. Maybe a user will manually send a request without `seq_no_primary_term` and will want to use our mapping code to map, or we've omitted some API that needs this to be specifically enabled.

rpuch commented:
I've removed static factories from `SeqNoPrimaryTerm`, made its constructor public and added assertions to the constructor.

The current batch seems to be closed. Please let me know if something is missng.

sothawo commented:
Thanks a lot for this contribution!

rpuch commented:
Thank you for your review and suggestions :)

sothawo commented:
you're welcome! Glad to have someone to provide such quality contributions.

benweet commented:
Nice contribution indeed.

@rpuch `SeqNoPrimaryTerm` doesn't seem to be updated once indexing succeeds (or am I missing something?). Could it be updated in [`AbstractElasticsearchTemplate.save()`](https://github.com/spring-projects/spring-data-elasticsearch/blob/master/src/main/java/org/springframework/data/elasticsearch/core/ReactiveElasticsearchTemplate.java#L211) like `populateIdIfNecessary()` does?

Thanks.

sothawo commented:
thanks for finding this @benweet. I created https://jira.spring.io/browse/DATAES-876 an issue for this
